File: 백준/Gold/2310. 어드벤처 게임/어드벤처 게임.java
import java.io.*;
import java.util.*;

public class Main {

    static int n = 0;


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        while(true) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            n = Integer.parseInt(st.nextToken());

            if(n == 0)
                break;

            char[] type = new char[n+1];
            int[] cost = new int[n+1];

            ArrayList<Integer> graph[] = new ArrayList[n+1];

            for (int i = 1; i <= n; i++) {
                graph[i] = new ArrayList<>();

                st = new StringTokenizer(br.readLine(), " ");
                char t = st.nextToken().charAt(0);
                type[i] = t;

                int c = Integer.parseInt(st.nextToken());
                cost[i] = c;

                while (st.hasMoreTokens()) {
                    int next = Integer.parseInt(st.nextToken());

                    if (next != 0)
                        graph[i].add(next);
                }
            }
            sb.append(search(type,cost,graph));
        }
        System.out.print(sb);

        br.close();
    }

    static String search(char[] t, int[] c, ArrayList<Integer>[] graph)
    {
        boolean[] v = new boolean[n+1];
        int[] dis = new int[n+1];

        for(int i=1;i<=n;i++)
            dis[i] = -1;

        PriorityQueue<room> pq = new PriorityQueue<>(
                (x,y) -> y.cost - x.cost
        );

        pq.add(new room(1,0));

        while(!pq.isEmpty())
        {
            room cur = pq.poll();

            if(t[cur.v] == 'E')
            {
                v[cur.v] = true;

                if(cur.v == n)
                    break;

                for(int next : graph[cur.v])
                {
                    if(dis[next] < cur.cost)
                    {
                        dis[next] = cur.cost;
                        pq.add(new room(next,dis[next]));
                    }
                }
            }

            else if(t[cur.v] == 'L')
            {
                int curC = cur.cost;

                if(curC <= c[cur.v])
                    curC = c[cur.v];

                v[cur.v] = true;

                if(cur.v == n)
                    break;

                for(int next : graph[cur.v])
                {
                    if(dis[next] < curC)
                    {
                        dis[next] = curC;
                        pq.add(new room(next,dis[next]));
                    }
                }
            }

            else{
                int curC = cur.cost;

                if(curC < c[cur.v]) {
                    continue;
                }

                v[cur.v] = true;

                if(cur.v == n)
                    break;

                for(int next : graph[cur.v])
                {
                    if(dis[next] < curC - c[cur.v])
                    {
                        dis[next] = curC - c[cur.v];
                        pq.add(new room(next,dis[next]));
                    }
                }
            }
        }

        if(v[n])
            return "Yes\n";
        else
            return "No\n";
    }
}
class room
{
    int v;
    int cost;

    room(int v,int cost)
    {
        this.v = v;
        this.cost = cost;
    }
}

File: 백준/Gold/2310. 어드벤처 게임/어드벤처 게임.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.collections.HashMap

var n = 0
val br = BufferedReader(InputStreamReader(System.`in`))
lateinit var st : StringTokenizer
lateinit var room : Array<Room>
lateinit var graph : Array<ArrayList<Int>>
fun main() {

    var sb = StringBuilder()

    while(init()) {
        room = Array(n+1){Room('E',0)}
        graph = Array(n+1){ArrayList<Int>()}

        for(i in 1..n) {
            st = StringTokenizer(br.readLine()," ")

            var w = st.nextToken()[0]
            var v = st.nextToken().toInt()

            room[i] = Room(w,v)

            while(true) {
                var nextV = st.nextToken().toInt()

                if(nextV == 0)
                    break

                graph[i].add(nextV)
            }
        }

        var tmp = if(search()) "Yes" else "No"
        sb.append(tmp).append("\n")
    }

    print(sb)

}

fun search() : Boolean{
    if(room[1].w == 'T')
        return false

    var visit = Array<Int>(n+1) {-1}

    var pq : PriorityQueue<MyNode> = PriorityQueue(
        {x,y -> y.v - x.v}
    )

    visit[1] = 0
    pq.add(MyNode(1,0 + if(room[1].w == 'E') room[1].v else 0))

    while(!pq.isEmpty()) {
        var cur = pq.poll()

        if(cur.v == n) {
            return true
        }

        for(next in graph[cur.v]) {
            if(room[next].w == 'E') {
                if(visit[next] < cur.c) {
                    visit[next] = cur.c
                    pq.add(MyNode(next,visit[next]))
                }
            }else if(room[next].w == 'L') {
                var nextC = if(cur.c > room[next].v) cur.c else room[next].v

                if(visit[next] < nextC) {
                    visit[next] = nextC
                    pq.add(MyNode(next,nextC))
                }
            } else {
                var nextC = cur.c - room[next].v

                if(nextC >= 0 && visit[next] < nextC) {
                    visit[next] = nextC
                    pq.add(MyNode(next,nextC))
                }
            }
        }
    }

    return false
}
data class MyNode (var v : Int, var c : Int)
data class Room (var w : Char, var v : Int)

fun init() : Boolean {
    n = br.readLine().toInt()

    if(n == 0)
        return false
    return true
}
